### PR TITLE
Include checklist task to migrate off h2

### PR DESCRIPTION
**Pull request by dpsutton**
New checklist item is shown when not hosted and app-db is h2. A divergence from others, this checklist is not shown if it is not triggered. In other words, we don't want to show this task as completed in the UI if they are hosted or using pg, we just want to omit the task completely. This is because we do not know if they upgraded or just started with pg/mysql or started in the cloud.

Did lots of cleanup to make this easier to grok and easier to test. Ultimately we don't have to redef a bunch of stuff to test the scenarios.

#### Implementation

The old checklist used to be a flat list that was grouped by the `:group`. And it was created by a multimethod

```clojure
(defmulti ^:private admin-checklist-entry
  {:arglists '([entry-name])}
  identity)

(defmethod admin-checklist-entry :add-a-database
  [_]
  {:title       (tru "Add a database")
   :group       (tru "Get connected")
   :description (tru "Connect to your data so your whole team can start to explore.")
   :link        "/admin/databases/create"
   :completed   (t2/exists? Database, :is_sample false)
   :triggered   :always})

...

(defn- admin-checklist-values []
  (map
   admin-checklist-entry
   [:add-a-database :set-up-email :set-slack-credentials :invite-team-members :hide-irrelevant-tables
    :organize-questions :create-metrics :create-segments]))
```

This made it hard to understand the final structure going over the wire, and led to duplicated db access (three different steps did `(t2/count Card)`).

Now the structure is a bit easier to see:

```clojure
(mu/defn ^:private checklist-items
  [info :- ChecklistState]
  (remove nil?
          [{:name  (tru "Get connected")
            :tasks (get-connected-tasks info)}
           (let [tasks (productionize-tasks info)]
             ;; don't show task to upgrade to mysql/pg if not triggered
             (when (-> tasks first :triggered)
               {:name  (tru "Productionize")
                :tasks (productionize-tasks info)}))
           {:name  (tru "Curate your data")
            :tasks (curate-tasks info)}]))
```

#### Screenshots

If on h2 and not hosted
image

And if hosted the task does not appear as completed, it is just omitted entirely:
image

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30126)

### CODE REVIEW

File: frontend/src/metabase/admin/settings/setup/components/SetupCheckList/SetupCheckList.jsx [hunk 1]
```diff
@@ -70,10 +72,26 @@ export const Task = ({ title, description, completed, link }) => (
...
-export default class SettingsSetupList extends Component {
+const TaskLink = ({ className, link, children }) => {
+  if (isSameOrSiteUrlOrigin(link)) {
+    return (
+      <Link className={className} to={link}>
+        {children}
+      </Link>
+    );
+  } else {
+    return (
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/admin/settings/setup/components/SetupCheckList/SetupCheckList.jsx]: @uladzimirdev fixed

**Comment by dpsutton** [comment]
has one failing test: `percy-visual-regression-tests`. Seems to be a flake that is being investigated (thread in `engineering-ci`)

File: test/metabase/api/setup_test.clj [hunk 2]
```diff
@@ -409,56 +407,136 @@
...
+                    (for [{group-name :name, tasks :tasks} (mt/user-http-request :crowberto :get 200 "setup/admin_checklist")]
+                      {:name  (str group-name)
+                       :tasks (for [task tasks]
+                                (-> (select-keys task [:title :completed :triggered :is_next_step])
+                                    (update :title str)))}))))
+    (testing "info about switching to postgres or mysql"
+      (testing "is included when h2 and not hosted"
+        (with-redefs [api.setup/state-for-checklist (constantly default-checklist-state)]
+          (let [checklist (mt/user-http-request :crowberto :get 200 "setup/admin_checklist")]
+            (is (= ["Get connected" "Productionize" "Curate your data"]
```
Comment (qnkhuat) [inline test/metabase/api/setup_test.clj:475]: not sure checking the group name is helpful in this case, I think we want to check the tasks title.

File: test/metabase/api/setup_test.clj [hunk 3]
```diff
@@ -409,56 +407,136 @@
...
-                        :completed    true
-                        :triggered    false
-                        :is_next_step false}]}]
-             (for [{group-name :name, tasks :tasks} (mt/user-http-request :crowberto :get 200 "setup/admin_checklist")]
-               {:name  (str group-name)
-                :tasks (for [task tasks]
-                         (-> (select-keys task [:title :completed :triggered :is_next_step])
-                             (update :title str)))}))))
+    (with-redefs [api.setup/state-for-checklist (constantly default-checklist-state)]
+      (is (partial= [{:name  "Get connected"
```
Comment (qnkhuat) [inline test/metabase/api/setup_test.clj:430]:
```suggestion
      (is (=? [{:name  "Get connected"
```

why not use the new method?

File: test/metabase/api/setup_test.clj (hunk 2 above)
Comment (dpsutton) [reply test/metabase/api/setup_test.clj:475]: the structure is
```clojure
[{:name "Get connected"
  :tasks [...]}
 {:name "Productionize"
  :tasks [...]}
 {:name "Curate your data"
  :tasks [...]}]
```

So the title of the group names goes hand in hand with the tasks they bring along. The "Productionize" group is omitted when the instance is hosted and that's all I'm testing here.

If that doesn't sway you I can get a set of all task titles and check that the migrate off of h2 task is not present. Interested to see if you think that's necessary as this feels fine to me.

**Review by markbastian — APPROVED** [review]
I don't have a lot of context on this feature, but nothing jumps out at me as an issue and +1 for unit tests.

**Pull request merged**